docs/cudf/source/developer.md

```diff
@@ -0,0 +1,280 @@
+  i.e., they are "views" into existing device memory allocations.
+  Inputs to libcudf functions are typically views,
+  although some libcudf functions also return return views.
+
+#### Step 1: Declare the function in `lib.pxd`
+
+To make the function callable from within Cython code,
+we must declare it in `lib.pxd`
+
+```cython
+# cudf/_libxx/lib.pxd
+
+cdef extern from "cudf/copying.hpp" namespace "cudf::experimental" nogil:
+    cdef unique_ptr[table] cpp_gather "cudf::experimental::gather" (
+        table_view source_table,
+        column_view gather_map,
+    )
+```
+
+Note that we omit the argument of type `rmm::mr::device_memory_resource*` in our Cython declaration.
+All libcudf APIs take this argument (optionally), and we currently just use the default value.
+
+Also note the use of
+[C name specification](https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#resolving-naming-conflicts-c-name-specifications)
+to disambiguate the name of the libcudf function with the Cython wrapper
+that we will write.
+
+#### Step 2: Add a Cython wrapper function
+
+Next, we create a Cython wrapper function that calls the libcudf function:
+
+```cython
+# cudf/_libxx/copying.pyx
+
+import pandas as pd
+
+from cudf._libxx.lib cimport *
+from cudf._libxx.column cimport Column
+from cudf._libxx.table cimport Table
+
+def _gather(Table source_table, Column gather_map):
+    assert pd.api.types.is_integer_dtype(gather_map.dtype)
+    cdef unique_ptr[table] c_result = (
+        cpp_gather(
+            source_table.view(),
+            gather_map.view()
+        )
+    )
+    return Table.from_unique_ptr(
+        move(c_result),
+        column_names=source_table._column_names,
+        index_names=source_table._index._column_names
+    )
+```
+
+The `.view()` methods of `Column` and `Table` respectively
+return a `cudf::column_view` and `cudf::table_view`.
+The `.from_unique_ptr()` staticmethods convert
+`unique_ptr[cudf::column]` and `unique_ptr[cudf::table]`
+int `Column` and `Table` respectively.
+Together, these set of functions ferry data back and forth
+between libcudf and cuDF.
+
+#### Step 3: Add corresponding method to the Frame class
+
+The next step is to implement the specific method in
+Frame/DataFrame/Series/Index.
+The `gather` operation applies to all `Frame` types,
+so it is implemented in the `Frame` class:
+
+```python
+# frame.py
+
+import cudf._libxx as libcudfxx
+
+class Frame(libcudfxx.Table):
+    """
+    Frame: A collection of Column objects with an optional index.
+
+    Parameters
+    ----------
+    data : OrderedColumnDict
+        An OrderedColumnDict mapping column names to Columns
+    index : Table
+        A Frame representing the (optional) index columns.
+    """
+
+    def _gather(self, gather_map):
+        if not pd.api.types.is_integer_dtype(gather_map.dtype):
+            gather_map = gather_map.astype("int32")
+        result = self.__clas_from_table(
```

Comment:
```suggestion
        result = self.__class_from_table(
```

Reply:
I edited your suggestion - hope you don't mind :)